lucet-module-data/src/signature.rs

```diff
@@ -0,0 +1,172 @@
+use crate::error::Error::{self, IOError, ModuleSignatureError};
+use crate::ModuleData;
+use byteorder::{ByteOrder, LittleEndian};
+pub use minisign::PublicKey;
+use minisign::{SignatureBones, SignatureBox};
+use object::*;
+use std::fs::{File, OpenOptions};
+use std::io::{self, Cursor, Read, Seek, SeekFrom, Write};
+use std::path::Path;
+
+pub struct ModuleSignature;
+
+impl ModuleSignature {
+    pub fn verify<P: AsRef<Path>>(
+        so_path: P,
+        pk: &PublicKey,
+        module_data: &ModuleData,
+    ) -> Result<(), Error> {
+        let signature_box: SignatureBox =
+            SignatureBones::from_bytes(&module_data.get_module_signature())
+                .map_err(|e| ModuleSignatureError(e))?
+                .into();
+
+        let mut raw_module_and_data =
+            RawModuleAndData::from_file(&so_path).map_err(|e| IOError(e))?;
+        let cleared_module_data_bin =
+            ModuleData::clear_module_signature(raw_module_and_data.module_data_bin())?;
+        raw_module_and_data.patch_module_data(&cleared_module_data_bin);
+
+        minisign::verify(
+            &pk,
+            &signature_box,
+            Cursor::new(&raw_module_and_data.obj_bin),
+            true,
+            false,
+        )
+        .map_err(|e| ModuleSignatureError(e))
+    }
+}
+
+struct SymbolData {
+    offset: usize,
+    len: usize,
+}
+
+pub struct RawModuleAndData {
+    pub obj_bin: Vec<u8>,
+    pub module_data_offset: usize,
+    pub module_data_len: usize,
+}
+
+impl RawModuleAndData {
+    pub fn from_file<P: AsRef<Path>>(path: P) -> Result<Self, io::Error> {
+        let mut obj_bin: Vec<u8> = Vec::new();
+        File::open(&path)?.read_to_end(&mut obj_bin)?;
+
+        let module_data_symbol_data = Self::symbol_data(&obj_bin, "lucet_module_data", true)?
+            .ok_or(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "module data not found",
+            ))?;
+        let module_data_len_symbol_data =
+            Self::symbol_data(&obj_bin, "lucet_module_data_len", true)?.ok_or(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "module length not found",
+            ))?;
+        assert_eq!(module_data_len_symbol_data.len, 4);
+        assert_eq!(
+            LittleEndian::read_u32(&obj_bin[module_data_len_symbol_data.offset..]) as usize,
+            module_data_symbol_data.len
+        );
+        Ok(RawModuleAndData {
+            obj_bin,
+            module_data_offset: module_data_symbol_data.offset,
+            module_data_len: module_data_symbol_data.len,
+        })
+    }
+
+    pub fn module_data_bin(&self) -> &[u8] {
+        &self.obj_bin[self.module_data_offset as usize
+            ..self.module_data_offset as usize + self.module_data_len]
```

Comment:
I only _just_ realized this as a rust-y replacement for `[a:b]` style indices for a > 0:
```
&self.obj_bin[self.module_data_offset as usize..][..self.module_data_len]
```
just a thought, this reads fine as-is